sentry_streams/sentry_streams/k8s/operator/handlers.py

```diff
@@ -0,0 +1,140 @@
+            pipelines[filename] = yaml.safe_load(config_txt)
+
+        return pipelines
+
+    def apply_pipelines(self, pipelines):
+        for pipeline_name, pipeline_config in pipelines.items():
+            self.logger.debug(f"Handling {pipeline_name=}")
+            cm = generate_configmap(
+                config=pipeline_config, configmap_template=self.configmap_template
+            )
+            apply_namespace(k8s_resources=[cm], namespace=self.NAMESPACE)
+
+            self.create_or_patch_configmap(body=cm)
+
+            deployments = generate_deployments(
+                config=pipeline_config,
+                deployment_template=self.deployment_template,
+                image=self.image,
+                container_name=self.container_name,
+            )
+            apply_namespace(k8s_resources=deployments, namespace=self.NAMESPACE)
+
+            for deployment in deployments:
+                self.create_or_patch_deployment(body=deployment)
+
+    def create_or_patch_configmap(self, body):
+        name, namespace = body["metadata"]["name"], body["metadata"]["namespace"]
+        try:
+            api_res = self.core_v1_client.read_namespaced_config_map(name=name, namespace=namespace)
+        except K8sApiException as exc:
+            if exc.status == 404:
+                # Create a new configmap.
+                self.logger.info(f"Create config map {name=}")
+                self.core_v1_client.create_namespaced_config_map(namespace=namespace, body=body)
+            else:
+                raise
+        # config-map already exists patch it
+        self.core_v1_client.patch_namespaced_config_map(name=name, namespace=namespace, body=body)
+
+    def create_or_patch_deployment(self, body):
+        name, namespace = body["metadata"]["name"], body["metadata"]["namespace"]
+        try:
+            api_res = self.apps_v1_client.read_namespaced_deployment(name=name, namespace=namespace)
+        except K8sApiException as exc:
+            if exc.status == 404:
+                self.logger.info(f"Create deployment {name=}")
+                self.apps_v1_client.create_namespaced_deployment(namespace=namespace, body=body)
+            else:
+                raise
+        # do not patch the image
+        for container in body["spec"]["template"]["spec"]["containers"]:
+            if container["name"] == self.container_name:
+                del container["image"]
+        self.apps_v1_client.patch_namespaced_deployment(name=name, namespace=namespace, body=body)
+
+    def register_handlers(self):
+        kopf.on.startup()(self.configure)
+        kopf.on.create("configmap", field="metadata.name", value=self.PIPELINES_CONFIGMAP)(
+            self.configure
+        )
+        kopf.on.update("configmap", field="metadata.name", value=self.PIPELINES_CONFIGMAP)(
+            self.configure
+        )
+
+
+streams_operator = StreamsOperator()
+streams_operator.register_handlers()
```

Comment:
Does kopf require the operator to be defined at module level ?
If not can we have it instantiated in the main function or in a function called by main passing the config parameters into the constructor rather than transparently taking them from envvars in the constructor itself? It makes the modularization a bit better and unit tests easier.